### PR TITLE
aws.sns - add kms related filter for sns and set-encryption action

### PR DESCRIPTION
**Pull request by thisisshi**
Adds related filter for sns:

```yaml
policies:
  - name: find-sns-topic-with-alias-name
    resource: sns
    filters:
      - type: kms-key
        key: c7n:AliasName
        value: alias/skunk/trails
```

Adds new action: set-encryption for sns

```yaml
policies:
  - name: set-sns-topic-encryption-default
    resource: sns
    actions:
      - type: set-encryption # by default enable encryption and key with alias/aws/sns

  - name: set-sns-topic-encryption
    resource: sns
    actions:
      - type: set-encryption
        key: alias/cmk/key # defaults to enable

  - name: unset-sns-topic-encryption
    resource: sns
    actions:
      - type: set-encryption
        enabled: False
```

### CODE REVIEW

File: tests/test_sns.py [hunk 1]
```diff
@@ -475,3 +475,113 @@ def test_sns_modify_add_and_remove_policy(self):
...
+                        'type': 'set-encryption'
+                    }
+                ]
+            },
+            session_factory=session_factory
+        )
+        resources = p.run()
+        self.assertEquals(len(resources), 1)
+
+        # disable encryption
```
Comment (kapilt) [inline tests/test_sns.py]: can you split these three scenarios out to separate tests.

Comment (thisisshi) [reply tests/test_sns.py]: done

File: c7n/resources/sns.py [hunk 2]
```diff
@@ -252,3 +255,69 @@ def process(self, resources):
...
+            key = ''
+
+        for r in resources:
+            try:
+                sns.set_topic_attributes(
+                    TopicArn=r['TopicArn'],
+                    AttributeName='KmsMasterKeyId',
+                    AttributeValue=key
+                )
+            except ClientError as e:
```
Comment (kapilt) [inline c7n/resources/sns.py]: blanket exceptions are sketchy, ideally this is catching specific exceptions and let the rest bubble up.

File: c7n/resources/sns.py [hunk 3]
```diff
@@ -252,3 +255,69 @@ def process(self, resources):
...
+    )
+
+    permissions = ('sns:SetTopicAttributes', 'kms:DescribeKey',)
+
+    def process(self, resources):
+        sns = local_session(self.manager.session_factory).client('sns')
+
+        if self.data.get('enabled', True):
+            key = self.data.get('key', 'alias/aws/sns')
+            if key != 'alias/aws/sns':
```
Comment (kapilt) [inline c7n/resources/sns.py]: so the service accepts 'alias/aws/sns' but needs an Arn for any other key?

Comment (thisisshi) [reply c7n/resources/sns.py]: actually no, I was testing with the incorrect alias which resulted in the key not being found. After some more testing it looks like you can supply alias/key, the key id, or the arn of the key. I'll be updating it to remove this as it's incorrect.

File: c7n/resources/sns.py (hunk 2 above)
Comment (thisisshi) [reply c7n/resources/sns.py]: originally i wanted to catch invalid inputs, e.g. setting an invalid kms key as the key but it looks like it doesn't do any validation on the key itself.

**Pull request merged**